source/time_stepping/RSF_convection_time_step.cc

```diff
@@ -0,0 +1,143 @@
+            fe_values_old.reinit (cell);
+            fe_values_old[this->introspection().extractors.velocities].get_function_values (this->get_old_solution(),
+                                                                                            velocity_values_old);
+
+            double max_local_velocity = 0;
+            double max_local_velocity_old = 0;
+            for (unsigned int q=0; q<n_q_points; ++q)
+              {
+                max_local_velocity = std::max (max_local_velocity,
+                                               velocity_values[q].norm());
+                max_local_velocity_old = std::max (max_local_velocity_old,
+                                                   velocity_values_old[q].norm());
+              }
+
+            if (this->get_parameters().include_melt_transport)
+              {
+                const FEValuesExtractors::Vector ex_u_f = this->introspection().variable("fluid velocity").extractor_vector();
+                fe_values[ex_u_f].get_function_values (this->get_solution(), fluid_velocity_values);
+                fe_values_old[ex_u_f].get_function_values (this->get_old_solution(), fluid_velocity_values_old);
+
+                for (unsigned int q=0; q<n_q_points; ++q)
+                  {
+                    max_local_velocity = std::max (max_local_velocity,
+                                                   fluid_velocity_values[q].norm());
+                    max_local_velocity_old = std::max (max_local_velocity_old,
+                                                       fluid_velocity_values_old[q].norm());
+                  }
+              }
+
+            max_local_speed_over_meshsize = std::max(max_local_speed_over_meshsize,
+                                                     max_local_velocity
+                                                     /
+                                                     cell->minimum_vertex_distance());
+            max_local_speed_over_meshsize_old = std::max(max_local_speed_over_meshsize_old,
+                                                         max_local_velocity_old
+                                                         /
+                                                         cell->minimum_vertex_distance());
+
+          }
+
+      const double max_global_speed_over_meshsize
+        = Utilities::MPI::max (max_local_speed_over_meshsize, this->get_mpi_communicator());
+      const double max_global_speed_over_meshsize_old
+        = Utilities::MPI::max (max_local_speed_over_meshsize_old, this->get_mpi_communicator());
+
+      double min_RSF_convection_timestep = std::numeric_limits<double>::max();
+
+      // if max velocities stayed the same or decreased, take the CFL number specified in the input file.
+      // else, take 0.5 as the CFL number to let the timestep increase faster
+      if ((max_global_speed_over_meshsize != 0.0) && (max_global_speed_over_meshsize <= max_global_speed_over_meshsize_old))
```

Comment:
I don't understand this logic. If the current velocity is smaller than the previous velocity, isn't that the moment when you want the timestep to increase faster?

Reply:
You are completely right, thanks for spotting this! I wrote it down for the timestep length, which becomes smaller when we want the smaller CFL number. And somehow I wrongly made it into a velocity rule...